src/fsharp/FSharp.Core/local.fs

```diff
@@ -1193,3 +1219,24 @@ module internal Array =
                 res.[i] <- subUnchecked !startIndex minChunkSize array
                 startIndex := !startIndex + minChunkSize
             res
+
+    let ofSeq (source : seq<'T>)  =
+        match source with
+        | :? ('T[]) as res -> (res.Clone() :?> 'T[])
+        | :? ('T list) as res -> List.toArray res
+        | :? ICollection<'T> as res ->
```

Comment:
This is another place where checking for the `IReadOnlyCollection<'T>` interface could be more efficient for some collections. Just like in `ThinEnumerable.Length`, you'd need to wrap that check in a `#if !FSCORE_PORTABLE_OLD` directive to ensure compatibility with older .Net versions (.Net Framework older than 4.5, Windows Phone older than 8.1, and so on). This would allow things like `FSharpx.Collection.PersistentVector`, which know their own length as an O(1) operation, to be converted directly to an array as well, avoiding the extra copy via ResizeArray. (Once PersistentVector acquires an implementation of `IReadOnlyCollection<'T>`, that is.)